### PR TITLE
Mv level work4

### PR DESCRIPTION
**Pull request by matthewvon**
4 small changes:  details here https://github.com/basho/leveldb/wiki/Mv-level-work4

### CODE REVIEW

File: db/version_set.h [hunk 1]
```diff
@@ -139,7 +139,7 @@ class Version {
   // are initialized by Finalize().
   double compaction_score_;
   int compaction_level_;
-  int write_penalty_;
+  volatile int write_penalty_;
```
Comment (gburd) [inline db/version_set.h:142]: I'm not sure what behavior you're trying to get with the addition of "volatile" as it doesn't prevent data race conditions or multi-instruction loads so frequently it doesn't accomplish what you'd think in multi-core/CPU-threaded code.  So, what's your intention here? (references: http://software.intel.com/en-us/blogs/2007/11/30/volatile-almost-useless-for-multi-threaded-programming http://stackoverflow.com/questions/8819095/concurrency-atomic-and-volatile-in-c11-memory-model/8833218#8833218 http://bartoszmilewski.com/2008/12/01/c-atomics-and-memory-ordering/)

File: util/env_posix.cc [hunk 2]
```diff
@@ -929,9 +930,15 @@ class PosixEnv : public Env {
     else
     {
         // low priority compaction (not imm, not level0)
+#if 1
```
Comment (gburd) [inline util/env_posix.cc]: Do you need to keep this if/else/endif around for some reason?  Seems to me to be just a reminder which could be a comment or simply left to the file's history in git.

File: util/env_posix.cc [hunk 3]
```diff
@@ -944,6 +951,30 @@ class PosixEnv : public Env {
  * support direct access to underlying type ... which is needed.
  */
 void
+PosixEnv::InsertQueue0(
+    PosixEnv::BGItem & item)
+{
+    BGQueue::iterator it;
+    bool looking;
+
+    // this is slow and painful with deque as underlying container
```
Comment (gburd) [inline util/env_posix.cc:954]: Indeed that does look slow/painful.  Looks like a priority queue insert into a dequeue data structure.  Does this happen on a path that's used frequently enough to bother looking for a better implementation?

**Comment by gburd** [comment]
With the exception of the confusing use of "volatile" I'm +1 for merge on this set of changes.  Even that seems minor given where it's used, but I'd like to clarify that first.

File: db/version_set.h (hunk 1 above)
Comment (matthewvon) [inline db/version_set.h:142]: (software.intel.com link worked yesterday, and failed today, so was not reread before writing this comment.  I think it has comments about what I am saying here.  The other two are not as relevant.)  Unless I made a mistake, no critical changes are ever made to a variable with volatile access without proper synchronization methods.  My intended usage of volatile is two fold.  First I like to use it to highlight variables to the next programmer that are referenced between threads independent of the the entire class / struct.  I may or may not have proper mutex or atomic synchronization on the variable depending upon its criticality.  Second I use volatile in hopes of limiting unexpected compiler optimizations that will change the code logic.  I am hinting to the compiler that the value should be reread.  The later is important if my intention is to reference the value of a variable that may be changing in another thread ... and make the reference without going through the expense of atomics / mutexes.

File: util/env_posix.cc (hunk 3 above)
Comment (matthewvon) [inline util/env_posix.cc:954]: The intent is to remove both InsertQueue0 and InsertQueue2.  The replacement will be a Set or Map which will keep contents in sorted order.  This kludge is to allow me to keep original Google container without a big rewrite.  The big rewrite is now overdue.  I have high hopes that is going to happen for Riak 2.0.

File: util/env_posix.cc (hunk 2 above)
Comment (matthewvon) [inline util/env_posix.cc]: Oops ... might be gone by the time you check the code again.

File: db/version_set.h (hunk 1 above)
Comment (gburd) [inline db/version_set.h:142]: In my experience and from what I've read the volatile keyword does more harm than good in most cases.  A comment could be used to indicate the intended use across threads.  But, I think this falls into the category of YMMV and isn't a critical concern.

**Comment by gburd** [comment]
+1 to merge, nice work

**Pull request merged**